File: src/eko-core/prompt/components/task.ts
import { ComponentConfig, ComponentContext, PromptComponent } from './types'
import { resolveTemplate } from '../template-engine'

/**
 * 任务上下文组件
 */
export const taskContextComponent: ComponentConfig = {
  id: PromptComponent.TASK_CONTEXT,
  name: 'Task Context',
  description: 'Task context information',
  required: false,
  template: `# Task Context
{taskContext}`,
  fn: async (context: ComponentContext) => {
    const { task, context: agentContext } = context

    if (!task && !agentContext?.chain.taskPrompt) return undefined

    const taskContext = task?.xml || agentContext?.chain.taskPrompt || ''
    if (!taskContext.trim()) return undefined

    const template = taskContextComponent.template!
    return resolveTemplate(template, { taskContext })
  },
}

/**
 * 任务节点组件
 */
export const taskNodesComponent: ComponentConfig = {
  id: PromptComponent.TASK_NODES,
  name: 'Task Nodes',
  description: 'Task node processing description',
  required: false,
  template: `# Task Node Processing
{nodePrompt}`,
  fn: async (context: ComponentContext) => {
    const { task } = context

    if (!task?.xml) return undefined

    return undefined
  },
}

/**
 * 任务示例组件
 */
export const taskExamplesComponent: ComponentConfig = {
  id: PromptComponent.TASK_EXAMPLES,
  name: 'Task Examples',
  description: 'Task processing examples',
  required: false,
  template: `# Task Processing Examples
{examples}`,
  fn: async () => {
    // 这里可以根据任务类型提供相应的示例
    // 暂时返回undefined，后续可以扩展
    return undefined
  },
}

File: src-tauri/src/agent/prompt/components/task.rs
use async_trait::async_trait;
use serde_json::json;
use std::collections::HashMap;
use std::sync::Arc;

use anyhow::{anyhow, Context};

use crate::agent::config::PromptComponent;
use crate::agent::error::AgentResult;
use crate::agent::prompt::components::types::{ComponentContext, ComponentDefinition};
use crate::agent::prompt::template_engine::TemplateEngine;

pub fn definitions() -> Vec<Arc<dyn ComponentDefinition>> {
    vec![
        Arc::new(TaskContextComponent),
        Arc::new(TaskNodesComponent),
        Arc::new(TaskExamplesComponent),
    ]
}

struct TaskContextComponent;

#[async_trait]
impl ComponentDefinition for TaskContextComponent {
    fn id(&self) -> PromptComponent {
        PromptComponent::TaskContext
    }

    fn name(&self) -> &str {
        "Task Context"
    }

    fn description(&self) -> &str {
        "Current task context and information"
    }

    fn required(&self) -> bool {
        false
    }

    fn dependencies(&self) -> &[PromptComponent] {
        &[]
    }

    fn default_template(&self) -> Option<&str> {
        Some("TASK CONTEXT\n\nUser Request: {user_prompt}\nTask Status: {task_status}\n{additional_context}")
    }

    async fn render(
        &self,
        context: &ComponentContext,
        template_override: Option<&str>,
    ) -> AgentResult<Option<String>> {
        let task = match &context.task {
            Some(task) => task,
            None => return Ok(None),
        };

        let template = template_override
            .or_else(|| self.default_template())
            .context("missing task context template")?;

        let additional_context = if context.additional_context.is_empty() {
            "".to_string()
        } else {
            format!(
                "\nAdditional Context:\n{}",
                serde_json::to_string_pretty(&context.additional_context)
                    .unwrap_or_else(|_| "{}".to_string())
            )
        };

        let mut template_context = HashMap::new();
        template_context.insert("user_prompt".to_string(), json!(task.user_prompt.clone()));
        template_context.insert("task_status".to_string(), json!(task.status.to_string()));
        template_context.insert("additional_context".to_string(), json!(additional_context));

        let result = TemplateEngine::new()
            .resolve(template, &template_context)
            .map_err(|e| anyhow!("failed to render task context template: {}", e))?;

        Ok(Some(result))
    }
}

struct TaskNodesComponent;

#[async_trait]
impl ComponentDefinition for TaskNodesComponent {
    fn id(&self) -> PromptComponent {
        PromptComponent::TaskNodes
    }

    fn name(&self) -> &str {
        "Task Nodes"
    }

    fn description(&self) -> &str {
        "Task node processing description"
    }

    fn required(&self) -> bool {
        false
    }

    fn dependencies(&self) -> &[PromptComponent] {
        &[]
    }

    fn default_template(&self) -> Option<&str> {
        None
    }

    async fn render(
        &self,
        _context: &ComponentContext,
        _template_override: Option<&str>,
    ) -> AgentResult<Option<String>> {
        Ok(None)
    }
}

struct TaskExamplesComponent;

#[async_trait]
impl ComponentDefinition for TaskExamplesComponent {
    fn id(&self) -> PromptComponent {
        PromptComponent::TaskExamples
    }

    fn name(&self) -> &str {
        "Task Examples"
    }

    fn description(&self) -> &str {
        "Task processing examples"
    }

    fn required(&self) -> bool {
        false
    }

    fn dependencies(&self) -> &[PromptComponent] {
        &[]
    }

    fn default_template(&self) -> Option<&str> {
        None
    }

    async fn render(
        &self,
        _context: &ComponentContext,
        _template_override: Option<&str>,
    ) -> AgentResult<Option<String>> {
        Ok(None)
    }
}
